### PR TITLE
[Improvement] Set RandAugment as Imgaug default transforms.

### PR DESCRIPTION
**Pull request by irvingzhang0512**
Use imgaug to reimplement [RandAugment](https://arxiv.org/abs/1909.13719).

According to [VideoMix](http://arxiv.org/abs/2012.03457), RandAugment helps a little.

image

## Results

+ sthv1 & tsm-r50, 8V100, 50epochs

|configs|top1 acc(efficient/accuracy)|top5 acc(efficient/accuracy)|
|:-|:-:|:-:|
|mmaction2 model zoo|45.58 / 47.70|75.02 / 76.12|
|testing with model zoo ckpt|45.47 / 47.55|74.56 / 75.79|
|training with default config|45.82 / 47.90|74.38 / 76.02|
|flip|47.10 / 48.51|75.02 / 76.12|
|randaugment|47.16 / 48.90|76.07 / 77.92|
|flip+randaugment|47.85/50.31|76.78/78.18|

+ Kinetics400, 8 V100, test with 256x256 & three crops

|Models|top1/5 accuracy|Training lost(epoch 100)|training time|
|:-:|:-:|:-:|:-:|
|TSN-R50-1x1x8-Vanilla|70.74%/89.37%|0.8|2days 12hours|
|TSN-R50-1x1x8-RandAugment|71.07%/89.40%|1.3|2days 22hours|
|I3D-R50-32x2x1-Vanilla|74.48%/91.62%|1.1|3days 10hours|
|I3D-R50-32x2x1-RandAugment|74.23%/91.45%|1.5|4days 10hours|

### CODE REVIEW

**Comment by irvingzhang0512** [comment]
@innerlee This pr is ready for review.
It would be great if someone could train kinetics-400 with/without RandAugment.

File: mmaction/datasets/pipelines/augmentations.py [hunk 1]
```diff
@@ -134,38 +134,58 @@ def __init__(self, transforms):
...
-                            dict(type='MedianBlur', k=(3, 11))
-                        ]),
+                        type='ShearX',
+                        shear=17.19 * cur_level * random.choice([-1, 1])),
+                    dict(
+                        type='ShearY',
+                        shear=17.19 * cur_level * random.choice([-1, 1])),
+                    dict(
+                        type='TranslateX',
+                        px=int(100 * cur_level * random.choice([-1, 1]))),
```
Comment (innerlee) [inline mmaction/datasets/pipelines/augmentations.py:167]: What's the assumption on input image size?

Comment (irvingzhang0512) [reply mmaction/datasets/pipelines/augmentations.py:167]: In official repo, when training ImageNet with RandAugment, the image sizes for ResNet/EfficientNet-B5/EfficientNet-B7 are 224/456/600.

Actually in other implementations of RandAugment, TranslateX/TranslateY use percentage params, such as 20%.

Thanks for pointing out.

Comment (innerlee) [reply mmaction/datasets/pipelines/augmentations.py:167]: Video frames are generally much smaller than image

Comment (irvingzhang0512) [reply mmaction/datasets/pipelines/augmentations.py]: Done, use `percent` instead of `px`

**Review by innerlee — APPROVED** [review]

**Comment by innerlee** [comment]
@congee524  would you like to do a comparison with this aug on/off on tsn&i3d?

**Comment by congee524** [comment]
> @congee524 would you like to do a comparison with this aug on/off on tsn&i3d?

oh, hh，I forgot about it. （ﾉ´д｀）

Because of the shortage of resources, it may start this Saturday.

**Comment by congee524** [comment]
#492

**Comment by irvingzhang0512** [comment]
I did some tests myself, with [tsn_r50_video_1x1x8_100e_kinetics400_rgb.py](https://github.com/open-mmlab/mmaction2/blob/master/configs/recognition/tsn/tsn_r50_video_1x1x8_100e_kinetics400_rgb.py)

Results(test with 256 ThreeCrop)

|Models|top1/5 accuracy|
|:-:|:-:|
|Vanilla|70.74%/89.37%|
|RandAugment|71.07%/89.40%|

**Comment by congee524** [comment]
> I did some tests myself, with [tsn_r50_video_1x1x8_100e_kinetics400_rgb.py](https://github.com/open-mmlab/mmaction2/blob/master/configs/recognition/tsn/tsn_r50_video_1x1x8_100e_kinetics400_rgb.py)
> 
> Results(test with 256 ThreeCrop)
> 
> Models	top1/5 accuracy
> Vanilla	70.74%/89.37%
> RandAugment	71.07%/89.40%

iφ(*￣0￣)， I‘m working on i3d

**Comment by irvingzhang0512** [comment]
> > I did some tests myself, with [tsn_r50_video_1x1x8_100e_kinetics400_rgb.py](https://github.com/open-mmlab/mmaction2/blob/master/configs/recognition/tsn/tsn_r50_video_1x1x8_100e_kinetics400_rgb.py)
> > Results(test with 256 ThreeCrop)
> > Models	top1/5 accuracy
> > Vanilla	70.74%/89.37%
> > RandAugment	71.07%/89.40%
> 
> iφ(*￣0￣)， I‘m working on i3d

Hope randaugment could help.
Unfortunately, the server must be shut down during the holidays(tears), otherwise I can train more models...

**Comment by congee524** [comment]
@irvingzhang0512 Hi, when I was training the ckpt of i3d with imgaug, I found that the training speed was very slow because the CPU was running at full load. Is this method CPU intensive?

**Comment by irvingzhang0512** [comment]
> @irvingzhang0512 Hi, when I was training the ckpt of i3d with imgaug, I found that the training speed was very slow because the CPU was running at full load. Is this method CPU intensive?

Yes. It takes me 3 days to train tsn + 1x1x8 + randaugment  with 8 V100 gpus

**Comment by irvingzhang0512** [comment]
Update some training results. RandAugment doesn't improve accuracy for I3D. Training more epochs may help.


|Models|top1/5 accuracy|Training lost(epoch 100)|training time|
|:-:|:-:|:-:|:-:|
|I3D-R50-32x2x1-Vanilla|74.48%/91.62%|1.1|3days 10hours|
|I3D-R50-32x2x1-RandAugment|74.23%/91.45%|1.5|4days 10hours|

**Comment by innerlee** [comment]
My bet is movie transition fx https://github.com/open-mmlab/mmaction2/issues/436#issuecomment-747166011 will be more helpful

**Comment by congee524** [comment]
> Update some training results. RandAugment doesn't improve accuracy for I3D. Training more epochs may help.
> 
> Models	top1/5 accuracy	Training lost(epoch 100)	training time
> I3D-R50-32x2x1-Vanilla	74.48%/91.62%	1.1	3days 10hours
> I3D-R50-32x2x1-RandAugment	74.23%/91.45%	1.5	4days 10hours

Could you provide your training log? I have trained 66 epoch Maybe the performance will be better

**Comment by irvingzhang0512** [comment]
[I3D_Vanilla.log](https://github.com/open-mmlab/mmaction2/files/6026616/I3D_Vanilla.log)
[I3D_RandAugment.log](https://github.com/open-mmlab/mmaction2/files/6026618/I3D_RandAugment.log)

FYI, there is a huge gap between our kinetics-400...

**Comment by congee524** [comment]
> [I3D_Vanilla.log](https://github.com/open-mmlab/mmaction2/files/6026616/I3D_Vanilla.log)
> [I3D_RandAugment.log](https://github.com/open-mmlab/mmaction2/files/6026618/I3D_RandAugment.log)
> 
> FYI, there is a huge gap between our kinetics-400...

It looks similar. Maybe I can stop the program :-P

**Comment by irvingzhang0512** [comment]
> It looks similar. Maybe I can stop the program :-P

Since our datasets are quite different, it's better to upload ckpts trained by your dataset.

**Comment by congee524** [comment]
> > It looks similar. Maybe I can stop the program :-P
> 
> Since our datasets are quite different, it's better to upload ckpts trained by your dataset.

OK, about 4 days left.

**Comment by congee524** [comment]
@irvingzhang0512 My result is 72.82%/90.64%. Maybe your ckpt should be uploaded.

**Comment by irvingzhang0512** [comment]
> @irvingzhang0512 My result is 72.82%/90.64%. Maybe your ckpt should be uploaded.

Worse than model without randaugment?
Since our datasets are different, my ckpt makes no sense in current mmaction2 model zoo.

**Comment by congee524** [comment]
> > @irvingzhang0512 My result is 72.82%/90.64%. Maybe your ckpt should be uploaded.
> 
> Worse than model without randaugment?
> Since our datasets are different, my ckpt makes no sense in current mmaction2 model zoo.

Yes, from 73.27 to 72.82.
Is the resolution of your dataset `short-side 320`? I use short-side 256.

**Comment by irvingzhang0512** [comment]
I use short-side 256. But our training&val samples(videos) are quite different.

https://github.com/open-mmlab/mmaction2/pull/415#issuecomment-768814184

**Comment by irvingzhang0512** [comment]
> Yes, from 73.27 to 72.82.
> Is the resolution of your dataset `short-side 320`? I use short-side 256.

I'll try randaugment in other datasets(sth) this month

**Comment by congee524** [comment]
> I use short-side 256. But our training&val samples(videos) are quite different.
> 
> [#415 (comment)](https://github.com/open-mmlab/mmaction2/pull/415#issuecomment-768814184)

ohh, I didn't use the default transforms, but used
```
dict(
        type='Imgaug',
        transforms=[
            dict(type='Fliplr', p=0.5),
            dict(type='Rotate', rotate=(-20, 20)),
            dict(type='Dropout', p=(0, 0.05))
        ]),
```

I'll train it again. sorry for my carelessness.

**Comment by irvingzhang0512** [comment]
+ training sthv2 with r50-tsm

|model|top1/5 accuracy for efficient mode|top1/5 accuracy for accurate mode|
|:-:|:-:|:-:|
|without randaugment(copy from model zoo)|57.86/84.67 |61.12/86.26|
|without randaugment(test with the [model zoo ckpt](https://download.openmmlab.com/mmaction/recognition/tsm/tsm_r50_1x1x8_50e_sthv2_rgb/tsm_r50_1x1x8_50e_sthv2_rgb_20200912-033c4ac6.pth) by myself)|55.56/82.94|56.92/83.92|
|with randaugment|58.83/84.75|61.42/86.56|

I test model zoo ckpt with [default config](https://github.com/open-mmlab/mmaction2/blob/master/configs/recognition/tsm/tsm_r50_1x1x8_50e_sthv2_rgb.py) except that I use videos as input instead of rawframes

**Comment by irvingzhang0512** [comment]
kindly ping @congee524 

training tsm-r50 on sthv1 with randaugment, get very good results. ckpt/json/log is [here](https://wp.cethik.com:443/link/858EA8B5B02FAB1D8231201F559A1BBF)

image


Maybe this ckpt is enough for this pr?

**Comment by congee524** [comment]
@irvingzhang0512 Sorry for my late reply...

When I use Imgaug with default transforms, The training time has increased a lot (about 18 days on 8 V100), which may be due to weak CPU performance. But the performance didn't improve (72.49/90.65).

**Comment by congee524** [comment]
Maybe Kinetics is large enough, so the image enhancement effect is very small, and simple `randmultiscale`, `flip` is enough.

For sthv1 and shtv2, since we didn't apply `flip` on them, `Imgaug` performs well.

I have found that the results of imgaug and flip with label transform on sthv1 are similar.

**Comment by irvingzhang0512** [comment]
> Maybe Kinetics is large enough, so the image enhancement effect is very small, and simple `randmultiscale`, `flip` is enough.
> 
> For sthv1 and shtv2, since we didn't apply `flip` on them, `Imgaug` performs well.
> 
> I have found that the results of imgaug and flip with label transform on sthv1 are similar.

Maybe we can add sthv1/tsm-r50/randaugment ckpt/log/json to model zoo instead of i3d-r50/kinetics/randaugment.

BTW, train sthv1 & tsm-r50 with flip & randaugment, get 47.85/50.31 top-1 accu1 and 76.78/78.18

**Comment by congee524** [comment]
I'll test your ckpt on our sthv1 now. (ง •_•)ง

**Comment by congee524** [comment]
> > Maybe Kinetics is large enough, so the image enhancement effect is very small, and simple `randmultiscale`, `flip` is enough.
> > For sthv1 and shtv2, since we didn't apply `flip` on them, `Imgaug` performs well.
> > I have found that the results of imgaug and flip with label transform on sthv1 are similar.
> 
> Maybe we can add sthv1/tsm-r50/randaugment ckpt/log/json to model zoo instead of i3d-r50/kinetics/randaugment.
> 
> BTW, train sthv1 & tsm-r50 with flip & randaugment, get 47.85/50.31 top-1 accu1 and 76.78/78.18

BTW, could you provide the ckpt of tsm_50_flip_randaugment ？

**Comment by irvingzhang0512** [comment]
https://wp.cethik.com:443/link/204D82F1146AF450E22C4E2BCEF67774

**Comment by congee524** [comment]
tsm_r50_randaugment on our dataset: 47.07/48.90 75.81/77.64

**Comment by irvingzhang0512** [comment]
I do know there is a gap between our sthv1 datasets, which is odd. We could directly download sthv1 rawframes from official website. So we are supposed to have exact the same dataset...

|configs|top1 acc(efficient/accuracy)|top5 acc(efficient/accuracy)|
|:-|:-:|:-:|
|mmaction2 model zoo|45.58 / 47.70|75.02 / 76.12|
|testing with model zoo ckpt|45.47 / 47.55|74.56 / 75.79|
|training with default config|45.82 / 47.90|74.38 / 76.02|

**Comment by congee524** [comment]
@dreamerlin may help to upload the ckpt/log/json :p

**Comment by congee524** [comment]
> https://wp.cethik.com:443/link/204D82F1146AF450E22C4E2BCEF67774

It seems not the tsm_50_flip_randaugment.

**Comment by irvingzhang0512** [comment]
> > https://wp.cethik.com:443/link/204D82F1146AF450E22C4E2BCEF67774
> 
> It seems not the tsm_50_flip_randaugment.

My mistake, 
https://wp.cethik.com:443/link/AB7BD549AF8760E17B9576A9C97C2C08

**Comment by congee524** [comment]
tsm_r50_flip_randaugment on our dataset:
47.75/50.16 76.52/77.89

File: configs/recognition/tsm/tsm_r50_flip_randaugment_1x1x8_50e_sthv1_rgb.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        filename_tmpl='{:05}.jpg',
+        pipeline=test_pipeline))
+evaluation = dict(
+    interval=2, metrics=['top_k_accuracy', 'mean_class_accuracy'])
+
+# optimizer
+optimizer = dict(weight_decay=0.0005)
+
+# runtime settings
+work_dir = './work_dirs/tsm_r50_flip_1x1x8_50e_sthv1_rgb/'
```
Comment (congee524) [inline configs/recognition/tsm/tsm_r50_flip_randaugment_1x1x8_50e_sthv1_rgb.py:98]: modify the work_dir

File: mmaction/datasets/pipelines/augmentations.py [hunk 3]
```diff
@@ -188,7 +208,10 @@ def imgaug_builder(self, cfg):
 
         obj_type = args.pop('type')
         if mmcv.is_str(obj_type):
-            obj_cls = getattr(iaa, obj_type)
+            try:
+                obj_cls = getattr(iaa, obj_type)
+            except AttributeError:
+                obj_cls = getattr(iaa.pillike, obj_type)
```
Comment (congee524) [inline mmaction/datasets/pipelines/augmentations.py:214]:
```python
obj_cls = getattr(iaa, obj_type) if hasattr(iaa, obj_type) else getattr(iaa.pillike, obj_type)
```

File: mmaction/datasets/pipelines/augmentations.py [hunk 4]
```diff
@@ -134,38 +134,58 @@ def __init__(self, transforms):
...
+        https://github.com/tensorflow/tpu/blob/master/models/official/efficientnet/autoaugment.py # noqa
+
+        Miss one augmenter ``SolarizeAdd`` since imgaug doesn't support this.
+
+        Returns:
+            dict: The constructed RandAugment transforms.
+        """
+        # RandAugment hyper params
+        n, m, max_level = 2, 9, 10
+        cur_level = 1.0 * m / max_level
```
Comment (congee524) [inline mmaction/datasets/pipelines/augmentations.py:152]: It seems that `n, m, max_level` is used only in some certain transform. Could you rename the variable with a more meaningful name?

**Comment by dreamerlin** [comment]
> > > https://wp.cethik.com:443/link/204D82F1146AF450E22C4E2BCEF67774
> > 
> > 
> > It seems not the tsm_50_flip_randaugment.
> 
> My mistake,
> randaugment: https://wp.cethik.com/#/link/858EA8B5B02FAB1D8231201F559A1BBF
> randaugment + flip: https://wp.cethik.com:443/link/AB7BD549AF8760E17B9576A9C97C2C08

uploaded

**Review by congee524 — APPROVED** [review]

**Comment by congee524** [comment]
@innerlee This PR is ready.

**Comment by innerlee** [comment]
Thanks!

**Pull request merged**